multi_partner_learning.py

```diff
@@ -0,0 +1,532 @@
+        self.single_partner_test_mode = single_partner_test_mode
+
+        # Attributes related to iterating at different levels
+        self.epoch_count = epoch_count
+        self.epoch_index = 0
+        self.minibatch_count = minibatch_count
+        self.minibatch_index = 0
+        self.is_early_stopping = is_early_stopping
+
+        # Attributes for storing intermediate artefacts and results
+        self.minibatched_x_train = [None] * self.partners_count
+        self.minibatched_y_train = [None] * self.partners_count
+        self.aggregation_weights = []
+        self.models_weights_list = [None] * self.partners_count
+        self.local_score_list = [None] * self.partners_count
+        self.score_matrix = np.zeros(shape=(self.epoch_count, self.partners_count))
+        self.score_matrix_extended = np.zeros(shape=(self.epoch_count, self.minibatch_count, self.partners_count))
+        self.global_val_acc, self.global_val_loss = [], []
+        self.is_save_fig = is_save_fig
+        self.save_folder = save_folder
+
+        logger.debug("MultiPartnerLearning object instantiated.")
+
+    def compute_test_score_for_single_partner(self, partner):
+        """Return the score on test data of a model trained on a single partner"""
+
+        logger.info(f"## Training and evaluating model on partner with id #{partner.id}")
+
+        # Initialize model
+        model = utils.generate_new_cnn_model()
+
+        # Set if early stopping if needed
+        cb = []
+        if self.is_early_stopping:
+            es = EarlyStopping(monitor='val_loss', mode='min', verbose=0, patience=constants.PATIENCE)
+            cb.append(es)
+
+        # Train model
+        logger.info("   Training model...")
+        history = model.fit(
+            partner.x_train,
+            partner.y_train,
+            batch_size=partner.batch_size,
+            epochs=self.epoch_count,
+            verbose=0,
+            validation_data=self.val_data,
+            callbacks=cb,
+        )
+
+        # Reference a testset according to the scenario configuration
+        if self.single_partner_test_mode == "global":
+            x_test, y_test = self.test_data
+        elif self.single_partner_test_mode == "local":
+            x_test, y_test = partner.x_test, partner.y_test
+        else:
+            raise NameError("single_partner_test_mode value '" + self.single_partner_test_mode + "' is not recognized.")
+
+        # Evaluate trained model
+        model_evaluation = model.evaluate(x_test, y_test, batch_size=constants.DEFAULT_BATCH_SIZE, verbose=0)
+        logger.info(f"   Model evaluation on test data: "
+                    f"{list(zip(model.metrics_names, ['%.3f' % elem for elem in model_evaluation]))}")
+
+        model_eval_score = model_evaluation[1]  # 0 is for the loss
+
+        # Return model score on test data
+        nb_epochs_done = (es.stopped_epoch + 1) if es.stopped_epoch != 0 else self.epoch_count
```

Comment:
I don't understand this line

Reply:
Via the callback the EarlyStopping provides `stopped_epoch`. But the way it is implemented, if early stopping wasn't trigger, this returns 0. So this line is a ternary operator for choosing between `stopped_epoch` and `epoch_count`.